src/Simplex_tree/include/gudhi/Simplex_tree.h

```diff
@@ -1958,21 +2028,24 @@ class Simplex_tree {
   }
 
  private:
-  bool rec_prune_above_filtration(Siblings* sib, Filtration_value filt) {
+  bool rec_prune_above_filtration(Siblings* sib, const Filtration_value& filt) {
     auto&& list = sib->members();
     bool modified = false;
     bool emptied = false;
     Simplex_handle last;
 
     auto to_remove = [this, filt](Dit_value_t& simplex) {
-      if (simplex.second.filtration() <= filt) return false;
-      if (has_children(&simplex)) rec_delete(simplex.second.children());
-      // dimension may need to be lowered
-      dimension_to_be_lowered_ = true;
-      return true;
+      //if filt and simplex.second.filtration() are non comparable, should return false.
```

Comment:
It was to match the multi-parameter case and I forgot somehow about the NaN case... If we disable this for multi-parameter instead, I would roll back the changes.

Reply:
I added a condition to includes the NaNs also. If you prefer that we disable this methods for any filtration value which is not an arithmetic native type, I will just put back the old conditions. Otherwise, I would add in the doc of `prune_above_filtration` that `operator<` is used to judge if a simplex should be removed or not.

Reply:
"arithmetic native type" is not the right condition. It is perfectly fine to use this function when the filtration value type is `CGAL::Lazy_exact_nt` or `mpq_class` or any other non-native scalar type.
The problematic case is multiparameter (which in the other branch can be tested with something like `Options::is_multi_parameter`), where the code will need heavier modifications (according to David's description of what the function should do, for instance), and thus it is safer to disable the function for now.
The main drawback of the old code is that it uses `<=` which isn't currently part of the concept (discussed in another comment).

Reply:
True, "arithmetic native type" is not the right condition, but I don't see how to discriminate the "right" classes from the "wrong" classes otherwise (I can imagine that someone can come up with other weird cases for filtration other than multi-peristence). In particular, if the reason we want to forbid multi-persistence is that the condition "<" is not always the right choice. "<" is not wrong per-se for multi-persistence, we are just not sure if it is the usual use case for it, if there is one.
So, I was just suggesting that we either explicitly forbid it for everythinh non trivial or that we allow it for everyone, but we highlight in the documentation, that "above" means "operator<" and if it is not what the user wants, he shouldn't use it.

Reply:
I also like solution 2 the best. Even though I was referring to `rec_prune_above_filtration` in the comment above which does not make use of intersect/unify. But the spirit is the same.

Reply:
I did this experiment https://gist.github.com/VincentRouvreau/2b406c867e3494eb3e4d4f50e566d50e on [bcdad85](https://github.com/GUDHI/gudhi-devel/pull/1122/commits/bcdad855d4d42319b4e65851716f225d6d74d23f) and it seems good to me.
Only serialize/deserialize would require some extra code for CGAL::Gmpq (serialize_trivial, deserialize_trivial, get_serialization_size_of and operator<<) to make it work properly.